### ISSUE TITLE
request: syntax for centering text

### ISSUE THREAD

**Issue by uvtc** [jgm/pandoc#719]
Now that Pandoc supports line blocks, here's some syntax that might work nicely for left/center/right -justifying lines:

```
Hi.

| left-justified
| preserves leading space
| last line

|< also left-justified
|< does not preserve leading space
|< last line

|> right-justified
|> another line
|> last line

|<> centered
|<> another line
|<> last line

Bye.
```

Thoughts?

**Comment by DaveJarvis** [comment]
Note that the following already has a meaning:

```markdown
| left-justified
| preserves leading space
| last line
```

There are two ways of marking up centered text that are already in the wild (different versions of markdown):

```
-> Centered Text <-
> Centered Text <
```

Leveraging the table syntax could be a way to force left/right justification:

```
->: Left Justified <-
-> Right Justified :<-
```

**Comment by Jmuccigr** [comment]
I would also have noted that the formatting, not semantic, super- and sub-scripting exist already in pandoc and in various markdown flavors.

**Comment by uvtc** [comment]
People will want to write their centered text in the center, like this:

~~~
                              -> some centered text <-

|<>                              some centered text
~~~

rather than

~~~
-> some centered text <-

|<> some centered text
~~~

because it just looks better that way.

But with the mirrored arrows syntax, how can you tell if it's meant to be centered text or verbatim text?

Another aspect of `|<>` I like is that since that marker is placed at column 0, it's easier to see if I've entered two or one newlines after (or before) typing in the centered text content (presuming I'm writing it centered in the page, as described above).

**Comment by DaveJarvis** [comment]
> People will want to write their centered text in the center, like this:

I prefer the `|<>` syntax to the mirrored arrows, so long as the following produce equivalent results:

```
|<> Centered Text
|<>                     Centered Text
```

|  How can you tell if [mirrored arrows are] meant to be centered text or verbatim text?

Good point. Also, some text editors might confuse `<-` with HTML comments (`<!--`), thus making a mess of syntax highlighting.

What are the next steps? Vote for what syntax to use, get agreement from the core developers, a prototype integration, or something else?

**Comment by DaveJarvis** [comment]
| I would also have noted that the formatting, not semantic, super- and sub-scripting exist already in pandoc and in various markdown flavors.

I actually considered it. Typed it in, then erased it as superfluous. Either way, you just did. ;-)

**Comment by linquize** [comment]
People like doing complex formatting,  where markdown cannot

**Comment by DaveJarvis** [comment]
> People like doing complex formatting

What "people"? That presupposes a certain type of people who want to use Markdown. Name them. Here are classes of people who could benefit from simple Markdown formatting:

* Technical writers
* Publishers
* Novelists
* Biographers
* Academics
* Bloggers

Here are classes of people who might require complex formatting, which makes Markdown unsuitable:

* Recipe authors
* Journalists
* Children's authors
* Screenplay writers (?)

> where markdown cannot

Nobody has drawn a line in the sand that states, "Markdown shall not do this." Markdown has no scope. Rather, we use subjective terms like "simple" and "human readable" to curb Markdown's extents. This will cause endless debate whenever additions are proposed.

The former planet Pluto was considered a planet because "planet" had no formal definition. Markdown suffers in the same way: it is has formal definition. That "Markdown cannot do this" doesn't mean "Markdown shall not do this."

**Comment by uvtc** [comment]
> What are the next steps? Vote for what syntax to use, ... {snip}

There's no voting, per se. From what I've seen, feature requests may be left open for a while to give folks time to mull them over, discuss, and possibly create patches. Of course, it's easy to add features but difficult to remove them (once folks are using them). Also, some requested features would be extremely useful, but might make the pandoc-markdown format too noisy, or clash with some other existing usage, or break backcompat, or not work with all (enough?) output formats.

> Nobody has drawn a line in the sand that states, "Markdown shall not do this." Markdown has no scope.

IMO, pandoc-markdown balances somewhere between capturing common usage (what you'd normally/naturally write in a plain text email or an online comment), and providing enough features to help you get away from Word/LaTeX/HTML. While looking good doing it. ;) A difficult task.

**Comment by tgy** [comment]
What is the current status for this? Markdown centering is something I would really like to see.

It is very important for scientific content when you want to center figures for example!

**Comment by ousia** [comment]
@tgy, as far as I remember, images accept attributes, so you can use attributes to center images.

I think it is way more important (and more useful) that paragraphs accept attributes than they may be aligned.

**Comment by v1kn** [comment]
I just found this issue, while looking for solutions to my problems. 
@DaveJarvis Thank You for a thorough essay on the rationale. I hope this feature will get implemented at some stage.

**Comment by ickc** [comment]
Probably this should be tagged "AST Change".

**Comment by v1kn** [comment]
@ickc 
It's quite a while since the issue was opened. Do You reckon there's possibility the Devs will work on it?

**Comment by ickc** [comment]
@v1kn,

I'm not the core-developers, so I can't tell. That said, I guess the priority will be low, considering that any "AST Change" level of issues are complicated and there are already many of them (in addition to the complexity involved, I would guess that a lot of thought would be given before any AST Change is made to get it right the first time). Just to illustrate the level of complexity: all reader/writer pairs for each format is needed to be updated, and depending on exactly what is involved, the tools built around pandoc needed to be updated too: pandoc-templates, pandoc-citeproc, any "filter framework" that parses the pandoc AST into its own, e.g. panflute, etc.

However, we can bend it such that it doesn't require an AST Change. e.g. just use a Div with class "center":

Cons:

- uglier, generic syntax (but Div might soon receive a markdown syntax, see #168, which also means even if we settled for a Div syntax for now, potentially in the future a markdown-ish syntax suggested above could be implemented in the future.)

- such a class is generic enough that existing documents might have already used it, so that such change might accidentally break someone else' documents when they update pandoc.

Pros:

- no AST Change, much easier to implement, and from the history of similar commits, it can be started with some common, easier to implement output formats first (e.g. LaTeX and HTML related outputs). i.e. not all formats has to be implemented at the same time.

- If one accept this route of using Div with class rather than a tailored syntax, then such feature can be implemented through a filter, completely bypass pandoc. It means you can start doing so right now, rather than waiting on a long discussion and approval process. Note that the filter as a proof of concept too, so it doesn't eliminate the possibility of pandoc supporting it natively.

**Comment by v1kn** [comment]
@ickc 
Thank You for a detailed explanation.  Yes I can see now this is no easy task with fast solution. 

I will try some workarounds using templates for the time being, but I will be sure to keep an eye on both this and #168  issue.

**Comment by jgm** [comment]
We might implement this at some point, but I'd advise using filters for now.
There are other AST changes, like colspans in tables, that are higher priority.

**Comment by DaveJarvis** [comment]
If issue #168 is resolved, then this issue can be closed, as an additional syntax for centering would be superfluous.

**Comment by ickc** [comment]
> If issue #168 is resolved, then this issue can be closed, as an additional syntax for centering would be superfluous

Not necessarily. It depends if one accepted the suggestion of using generic Div and a class to centering, or want a dedicated syntax for centering, similar to LineBlock (the original suggestions make comparison to LineBlock, and consider centering/LineBlock to be special cases of justifications).

In terms of AST, while one could imagine LineBlock syntax can be parsed as a Div with a certain class, but pandoc actually assigned a special element to it, which by the way cannot carries attributes at the moment. So to solve the centering problem in terms of AST, one can tries at least these methods:

1. dedicated centering element (how about right justified then?)
2. A Div with class to indicate it is centering
3. Overload the LineBlock element, and let it carries attributes to indicate if it is left-justified (the old LineBlock), centered, or right-justified.

So when I suggest using a Div, it is the 2nd approach. But the 3rd approach could as well be a viable option, especially if the syntax of centering will be similar to LineBlock's.

So there're 2 related considerations: 

1. philosophically, does one want to generalize LineBlock to include centering/right-justification? (i.e. how one categories it.)

2. syntactically, does one want to have a dedicated syntax for centering/right-justification? And if so, does one want to make them similar to LineBlocks?

If the answers are no to both, then using Div with classes is the simplest solution, and is a solution that can happens right now.

Lastly, I have a related question: what if one want to center/right-justified other kind of elements, say Headings? If this is desirable, then using a class can immediately generalize it to any pandoc elements that can carries attributes, while overloading LineBlock will limit it to just LineBlock-like texts.

**Comment by DaveJarvis** [comment]
> if one accepted the suggestion of using generic Div and a class to centering

Yes, this is the reason I included an example block, to show the syntax that would obviate this feature request:

```
; --- div {.centered}
...
```

> what if one want to center/right-justified other kind of elements, say Headings

Headings and other elements typically get their own "class" that can be styled without a specific class attribute. This applies equally to HTML and TeX code. It's even possible to style the same heading level in different ways:

```
# First Heading
Paragraph
# Second Heading
Paragraph
```

Using CSS3:

```
h1:nth-of-type(2) {
    text-align: center;
}
```

**Comment by ickc** [comment]
@DaveJarvis,

```markdown
; --- div {.centered}
...
```

compared to @uvtc's for example,

```markdown
|<>               centered
|<>       also ignores leading space
|<>               last line
```

The latter is definitely better fitting the philosophy:

> A Markdown-formatted document should be publishable as-is, as plain text, without looking like it’s been marked up with tags or formatting instructions.

I remembered @jgm mentioned somewhere (probably in CommonMark when commenting on Microsoft's another attempt on a tool similar to pandoc/MultiMarkdown) that using English words to describe what it is is worse than using a syntax to show it (e.g. non-English documents). While we are talking about a class here (not just some syntax using English word), when someone read it as a plain-text, it is the same to them. i.e. all they read is just an English word.

Just to make it clear, I am not opposing to either approaches (that's why I suggested using a Div instead earlier). But I'm just pointing out the others might have other considerations. So we need to wait if they have any other opinions before closing it.

And even if one settled to use classes for centering, should one do it with Div, or LineBlock (LineBlock cannot carries attributes right now, but it seems that it will eventually be supported)? That's the "philosophical" part I mentioned above.

**Comment by DaveJarvis** [comment]
> LineBlock cannot carries attributes right now, but it seems that it will eventually be supported?

I understand, thank you for clarifying. 

> Readability, however, is emphasized above all else. A Markdown-formatted document should be publishable as-is, as plain text, without looking like it’s been marked up with tags or formatting instructions.

That is an issue. Perhaps @uvtc, @jgm, or @jgruber might want to chime in with their thoughts on this potential divergence from Markdown philosophy.

**Comment by jgm** [comment]
+++ ickc [Jan 07 17 12:26 ]:
>   I remembered @jgm mentioned somewhere (probably in CommonMark when
>   commenting on Microsoft's another attempt on a tool similar to
>   pandoc/MultiMarkdown) that using English words to describe what it is
>   is worse than using a syntax to show it (e.g. non-English documents).
>   While we are talking about a class here (not just some syntax using
>   English word), when someone read it as a plain-text, it is the same to
>   them. i.e. all they read is just an English word.

My main objection to the use of English words is that not
everyone writes in English.

One nice thing about Markdown as opposed to, say, LaTeX, is
that if you're writing in (say) Swedish, your document isn't
littered with English words.  It's ALL in Swedish, with some
punctuation.  People have sometimes commented on how nice
this is.

**Comment by DaveJarvis** [comment]
```
; --- {.poesi}
...
; --- {.poem}
...
```

How the "poem" is presented (centered, right-justified, etc.) is independent of the Markdown, as is the language used to denote that the paragraph is a stanza. That is to say, no formatting instructions are present, though `{.poem}` could be considered a tag, which goes against the philosophy.

**Comment by bpj** [comment]
I came across this when looking for line block related issues to check if another issue I have had been reported.  I have not read everything here (it's a lot!) but I thought that now when we have a nice div syntax what about making a class like `.center` (and perhaps also `.centre` to prevent problems due to human errors) on a div magical, so that writers output the right format dependent markup around it if possible?

In LaTeX that would presumably be the `center` environment, or a `minipage` with `centering` at the top (see [here](https://tex.stackexchange.com/a/23653/93534) for the difference!)

I think that in HTML it should perhaps be up to the user to implement/link the necessary CSS based on the class rather than polluting the default template with more in-document CSS.  Remember that it's hard or impossible to override in-document CSS, whether in the header or in a `style` attribute!  The usually wanted CSS `.center { text-align: center; }` could be mentioned in the documentation of the feature.

**Comment by igormcoelho** [comment]
Any updates on this? (how to center text on LaTeX via pandoc)

I wonder if these fenced_divs can be used somehow: https://github.com/jgm/pandoc/issues/4037
Something link:
```
::: center
             test
:::
```

Or pehaps this pandoc-latex-environment, to somehow convert a div class "center" to the \begin{center} ... on LaTeX: https://github.com/chdemko/pandoc-latex-environment/wiki
```
---                           
pandoc-latex-environment:
  center: [classcenter]
---
<div class="classcenter">content</div>

in LaTeX ->

\begin{center}
content
\end{center}
```

Sorry, I don't know much of the internals of pandoc (yet), and couldn't make any of them work... can someone give a clue on the latest advances here?

**Comment by DaveJarvis** [comment]
@igormcoelho, you can create a simple Lua script to convert custom blocks to centred text. Here's an example that creates a custom inline image based on a fenced div:

```
local lines_to_blocks = {
  Image = function( el )
    return {
      pandoc.RawInline( "tex", "\\inlineexternalfigure[" .. el.src .. "]" )
    }
  end
}

function Div( el )
  local kls, _ = el.classes:find_if(
    function ( c )
      return string.match( c, "^image%-" )
    end
  )

  if kls then
    return pandoc.walk_block( el, lines_to_blocks )
  end
end
```

See also @jgm's https://github.com/jgm/pandoc/issues/2106#issuecomment-371508848.

Save the file as `centre.lua` then call pandoc using:

    pandoc --lua-filter=centre.lua

This then parses the following block:

```
::: image-inline
![](../climate/graph)
:::
```

To produce:

    \inlineexternalfigure[../climate/graph]

Feel free to use this example to produce centred LaTeX code. Note that `RawInline` may have to be changed to `RawBlock`.

**Comment by igormcoelho** [comment]
Thanks @DaveJarvis for the fast and detailed response.
I can try to adapt this code, but my need involves centering an equation block on LaTeX, not an image... in fact, I want to do that in a way that is compatible with markdown too, because we have users generating both LaTeX output and also markdown. Anyway, I'll try to change your script to `RawBlock` and learn a little bit of lua :)

**Comment by sjackman** [comment]
`$$maximize \sum_{i \in X} p_{i}$$` is centered in PDF output, but not in HTML output.

**Comment by igormcoelho** [comment]
> $$maximize \sum_{i \in X} p_{i}$$ is centered in PDF output, but not in HTML output.

I see that, problem is that I'm using (@obj) to create a label that will be further cited:
```
(@obj) $maximize \sum_{i \in X} p_{i}$

On Eq. (@obj) bla bla bla...
```
I think that double $$ environment in LaTeX do not allow labels...
And anyway, I have other parts of the text that I also need to centralize, this was just an example, so I've looking for a general way to do that on pandoc (since it's already possible on pandoc html by using div + css syntax).

**Comment by sjackman** [comment]
Have you tried using `pandoc -Fpandoc-crossref`?
```
$$ math $$ {#eq:label}
```
See http://lierdakil.github.io/pandoc-crossref/#equation-labels

**Comment by igormcoelho** [comment]
@sjackman thanks a lot for this advice, besides solving my equation centralization issue, pandoc-crossref improved a lot the general readability of the text (removed all my \ref{bla} to a much more friendly @ref). Thanks!

**Comment by sjackman** [comment]
I'm glad to hear that it helped!

**Comment by uvtc** [comment]
If my suggestion (here, above, from 2013!) to build upon line block syntax is not desirable (possibly because adding a three-character `|<>` side-mark syntax looks a lot like markup, and also may be going down a slippery slope of new multi-character syntax elements), consider the following:

  * Any centered syntax should look like what it means. It should look good and not too much like markup.
        
  * Centering text is actually pretty commonly needed, so the syntax for it should ideally be fairly succinct. In general, markdown avoids big multi-character markers anyway.

  * Markdown provides numerous ways to create a horizontal rule (HR). It's not *that* common to need an HR, but when you need one, there's a lot of ways to get one, and markdown doesn't care which one you use. This actually uses up some characters we might've preferred to use for something else, but here we are.
    
  * Typically, HR's are drawn centered.
    
  * The essence of an HR is that it's a divider; a delineation and a marker. Centered text is also a delineation and a marker of sorts.

  * I don't think we can re-use the fenced div syntax,
  
        :::             centered text!
    
    because that looks like the start of a fenced div block. And anyway, markdown tends to use different markers for fenced syntax vs side-marked syntax for a given feature (for example, code blocks are prepended with spaces, but fenced with tildes or backticks).
    
  * An enclosing syntax could maybe work, but we're low on matching bracket characters, and so it would have to be some compound syntax, for example:
  
        [- some centered text? -]
        
    but that looks too much like markup, and I doubt people want to go down that road of a slew of future compound marker proposals like that. And anyhow, that enclosing syntax is associated with inline markup, not block markup.
    
  * And, of course, you've got to have some marker on the far left-hand-side, even for centered text, otherwise it'll be mistaken for an indented code block. But the marker on the left-hand-side isn't so bad, because it helps you see which line the centered text is on.
    
Give all that, how about this simple rule: if you follow a horizontal rule marker with spaces followed by text, then the text gets centered and the horizontal rule itself is omitted.

~~~
---                       here is some centered text
~~~

That would be for single-line centered text only, though I suppose could also support a "lazy-style" (multi-line) variant.

> Note: it's not a typical side-marked syntax (like blockquote, which encourages multiple lines, and which often uses a vertically-oriented character marker), but rather it's an extension of the existing HR syntax.

Does that cause ambiguities with any other pandoc-markdown syntax? I think it's pretty clear it's not an HR. Does it break anything? The only possible issue that comes to my mind is that `---` usually gives me an em-dash.

~~~
_ _ _                          \* \* \*
~~~

So, this is a simple solution that's:

  * easily recognizable/readable,
  * visually distinct from an HR,
  * doesn't require any new syntax markers (and actually uses some that were previously locked up and unavailable for mardown syntax since they were exclusively used for HR's),
  * shouldn't render badly on impl's that don't support it, and
  * pretty well looks like what it means.

What do you think?

**Comment by jgm** [comment]
John Gabriele <notifications@github.com> writes:

> Does that cause ambiguities with any other pandoc-markdown syntax? I think it's pretty clear it's not an HR. Does it break anything? The only possible issue that comes to my mind is that `---` usually gives me an em-dash.


Well, in addition,

    * * * *      hi

currently gives you a nested list in pandoc and most other
markdown implementations.

**Comment by sjackman** [comment]
Perhaps some variation of the table syntax that renders a paragraph rather than a table.

```
|:---:|
A paragraph of centered text goes here.

|----:|
A paragraph of right-aligned text goes here.
```

**Comment by uvtc** [comment]
@jgm Ah, right. Thanks for pointing that out. I see there's the same issue with `- - -` as well.

It's difficult to come up with light, sensible, obvious markup for centered text, which also degrades gracefully (for implementations that don't support it) and at the same time looks "classic" such that it will be readily adopted (begins being used not only because it's handy but because it just "looks right").

I think the difficulty is partly because most folks would, in an email, probably just write their "centered text" indented with spaces and be done with it.

                             some centered text

@sjackman , an issue with re-using table syntax is that, aside from possible ambiguity with table syntax, you are also then back to using multi-character markers (`|:--` and so on), which could turn into an asciibet soup of new markers to remember. (I like `|<>`, `|>`, `|<` because they kinda generalize line block syntax, but I understand that consistency isn't everything; brevity, simplicity, and good looks probably beat consistency here.)

Remember that the *primary* table syntax in pandoc-markdown is so simple that you may not even realize you're using it. :) The other table syntaxes (like `|:--|`) are provided to give you more specialized formatting options, and/or are more easily generated by computer, but they're less "markdownish" and look more like markup.

**Updated tightened-up suggestion:** If a horizontal rule marker *of the kind containing no interspersed spaces* is followed by text, then the text is centered and the HR marker itself is omitted.

~~~
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
minim veniam, quis nostrud exercitation ullamco laboris nisi ut
aliquip ex ea commodo consequat. Duis aute irure dolor in
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
culpa qui officia deserunt mollit anim id est laborum.

--------                 some centered text

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
minim veniam, quis nostrud exercitation ullamco laboris nisi ut
aliquip ex ea commodo consequat. Duis aute irure dolor in
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
culpa qui officia deserunt mollit anim id est laborum.

********                 some centered text

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
minim veniam, quis nostrud exercitation ullamco laboris nisi ut
aliquip ex ea commodo consequat. Duis aute irure dolor in
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
culpa qui officia deserunt mollit anim id est laborum.

________                 some centered text

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad
minim veniam, quis nostrud exercitation ullamco laboris nisi ut
aliquip ex ea commodo consequat. Duis aute irure dolor in
reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
culpa qui officia deserunt mollit anim id est laborum.
~~~

This is simple, looks good, and seems obvious that it's meant to indicate centered text. It feels markdownish:

  * it's flexible (you may use more than 3 marker characters if you like, and you get to pick the marker character --- just like with HR's)
  * the characters are just all the same single character but repeated (no complicated character combos)
  * it's possible to (in the future?) support a lazy variant

    ~~~
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
    eiusmod tempor incididunt ut labore et dolore magna aliqua.

    ******               some centered text
                         with a second line

    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
    eiusmod tempor incididunt ut labore et dolore magna aliqua.
    ~~~

And I think, due to the similarities between between HR's and centered text, it works pretty well.

**Comment by mb21** [comment]
> Centering is a presentational instruction, and Markdown should separate semantics from presentation

I tend to agree with this. That's why I'm not so sure we need new syntax. Following the precedent of [smallcaps](http://pandoc.org/MANUAL.html#small-caps), underline and [columns](http://pandoc.org/MANUAL.html#columns), I would be fine with using native div and span syntax and use something like the following:

```
::: centered :::
my text
:::
```

Currently, you need to write a filter to handle this, but this could potentially be handled by pandoc itself, if we decide it's common enough.

**Comment by DaveJarvis** [comment]
The following mixes content with presentation:

```markdown
::: centered :::
my text
:::
```

As does:

```markdown
::: centered small
Copyright notice
:::
```

Consider, instead:

```markdown
::: footer
Copyright notice
:::
```

That is, rather than using the word "centered" to denote centered text, perhaps emphasis can be placed on separating content from presentation? Why does the text need to be centered? Is it a quotation? An indicator that a line of reasoning has stopped? A stanza from a poem? A header? A call-out? A warning?

**Comment by joshhansen** [comment]
Couldn't this be achieved using a heuristic instead of adding new syntax? Here's one approach:

For each line in a line block, we determine alignment based on the number of leading space characters, and the implied number of trailing space characters.

Leading spaces are the number of spaces between the pipe and the first non-space character (if any).

Implied trailing space characters are equal to the max line length in the block minus the position of a given line's last non-space character.

For example, consider this line block:
```
|left1
| left2
|  centered
|       right
```
In this block, the max line length is 13 ("right"). For each line in this block, we have the following number of leading and implied trailing spaces:
```
left1:    leading=0, trailing=7
left2:    leading=1, trailing=6
centered: leading=2, trailing=2
right:    leading=7, trailing=0
```
For each line, the alignment is determined as follows:
* if trailing=0 and leading > 1: align right
* if leading > 0 and trailing > 0 and leading approximately equals trailing: align center
* otherwise align left

This would give each line in the line block the expected alignment, using only the most natural of formatting. The main tricky bit seems to be defining "approximately equal" for determining if there are roughly the same number of leading and trailing spaces.

It would admittedly be a bit painful to manually align your text to the right, but it would also be what you'd have to do to get a nice-looking right-alignment in plain text. Perhaps editor macros/plugins could help with the task?

Thoughts? I'm not a Markdown expert so I'd guess I'm missing something obvious....

**Comment by alerque** [comment]
@joshhansen I'm sorry, but I think that particular suggestion fails on two counts:

1. Markdown traditionally tries to ignore quantities of space as much as possible. The existing variants to this (such as the 1–4 space indents to  keep content connected to list items) are already problematic across implementations. This is likely to lead to both confusion and incompatibility.

2. The line-block syntax in particular is special in that it typically _preserves_ leading spaces. This is often used for poetry markup. E.g.:

    ```markdown
    | Once a line
    |     indent me.
    | Twice over
    |     me too.
    ```

    ...will convert to LaTeX as follows:

    ```tex
    Once a line!\\
    \hspace*{0.333em}\hspace*{0.333em}\hspace*{0.333em}\hspace*{0.333em}indent me.!\\
    Twice over!\\
    \hspace*{0.333em}\hspace*{0.333em}\hspace*{0.333em}\hspace*{0.333em}me too.
    ```

    This behavior just happens to be critical to the book publishing I do using Markdown as the canonical content source. Trying to change the existing  behavior would screw me and a lot of other people up.

**Comment by joshhansen** [comment]
Good point. This could be addressed by only considering explicitly present trailing spaces rather than implied trailing spaces (implied by a line being shorter than the maximum line length in the block).

Consider this version (note the trailing spaces after "centered"):
```
|left1
| left2
|  left___3
|  centered  
|       right
```

The alignment rule then becomes:
```js
function alignment(line_length, max_line_length, leading, trailing) {
    if(line_length == max_line_length && leading > 1) {
        if(trailing == 0) {
            return "right";
        else if(leading approximately equals trailing) {
            return "center";
        }
    }
    return "left";
}
```

I'm not sure what you mean by "a centered paragraph with wrapped lines". Can you please expound?

Of course, as with many other aspects of pandoc, this would probably only make sense as an extension that's off by default. You mentioned poetry---that's my current use case, and I would find an extension like this incredibly useful.

**Comment by ftomassetti** [comment]
Is this still being considered?

**Comment by blaggacao** [comment]
```
CenTerD>TEXT  [WRITE TEXT HERE]
```
Jokes aside:
```
------       centered text
```
```
------       THE END OF #719
```

**Comment by alpianon** [comment]
> > Centering is a presentational instruction, and Markdown should separate semantics from presentation
> 
> I tend to agree with this. That's why I'm not so sure we need new syntax. Following the precedent of [smallcaps](http://pandoc.org/MANUAL.html#small-caps), underline and [columns](http://pandoc.org/MANUAL.html#columns), I would be fine with using native div and span syntax and use something like the following:
> 
> ```
> ::: centered :::
> my text
> :::
> ```
> 
> Currently, you need to write a filter to handle this, but this could potentially be handled by pandoc itself, if we decide it's common enough.

A simple filter could be this:

``` lua
function Div(el)
  local class = el["c"][1][2][1]
  if class == "center" or class == "right" then
    local align = class
    el["c"][1][2] = {} -- remove class
    el["c"][1][3] = {
      { "style", "text-align:"..align..";" }, { "custom-style", align }
    }
    --[[ add attributes both for html and for docx (you need to define "center"
         and "right" custom styles in your reference docx file) ]]
    return el
  end
  return nil
end
```

It works with conversion to both html and docx, but to make it work with docx you need to define "center" and "right" custom styles in your reference docx file.

If you start the text line with a `\`, you can also render the alignment in your markdown text, if you wish:

``` markdown

::: right
\                                                                                  this text is right-aligned
:::

::: center
\                                                  centered!
:::

```

**Comment by fbertoia** [comment]
Hello to all,
No update for centering text ? 😭

**Comment by bpj** [comment]
After all it is very simple to make

``````
::: {.center}
Centered.
:::
``````

work:

``````lua
local center_for = {
  latex = {
    pre = pandoc.RawBlock('latex', '\\begin{center}'),
    post = pandoc.RawBlock('latex', '\\end{center}'),
  },
  -- add more as needed...
}

function Div (div)
  if div.classes:includes('center') then
    if center_for[FORMAT] then
      local rv = {}
      if center_for[FORMAT].pre then
        rv[#rv+1] = center_for[FORMAT].pre
      end
      rv[#rv+1] = div
      if center_for[FORMAT].post then
        rv[#rv+1] = center_for[FORMAT].post
      end
      return rv
    end
  end
  return nil
end
``````

Not even 25 lines, and then I even made it customizeable.

**Comment by mdengler** [comment]
This would definitely be nice to have

**Comment by Cheetoacidosis** [comment]
God, this has been open for over a decade. ffs, add centering already!!

**Comment by raffaem** [comment]
I have found this series of filter in the official page for new pandoc filters: https://github.com/pandoc-ext/fonts-and-alignment

**Comment by mdengler** [comment]
Is status:more-discussion-needed resolvable?  The solutions seem to be:

1. hr-line leading `=========` etc., by @uvtc in https://github.com/jgm/pandoc/issues/719#issuecomment-469087684
2. `::: center` fences

...though #2 has disavantages (english word mentioned by @jgm earlier, and confusion with fenced divs mentioned by @uvtc .

So what about the `========` syntax?  Can we end the discussion @jgm?

**Comment by cagix** [comment]
> Is status:more-discussion-needed resolvable? The solutions seem to be:
> 
> 1. hr-line leading `=========` etc., by [@uvtc](https://github.com/uvtc) in [request: syntax for centering text #719 (comment)](https://github.com/jgm/pandoc/issues/719#issuecomment-469087684)
> 2. `::: center` fences
> 
> ...though [#2](https://github.com/jgm/pandoc/issues/2) has disavantages (english word mentioned by [@jgm](https://github.com/jgm) earlier, and confusion with fenced divs mentioned by [@uvtc](https://github.com/uvtc) .
> 
> So what about the `========` syntax? Can we end the discussion [@jgm](https://github.com/jgm)?

Why do we need a special syntax just to center text? Since Pandoc supports fenced divs, couldn't we just use the existing Pandoc Markdown syntax of `::: center ... :::`?

The simplest way would be for Pandoc to treat all fenced divs, such as `::: foobar`, in the same way, i.e. initially incorporating the properties into the ast w/o any further transformation. Writers need to emit the corresponding output, e.g. for HTML setting the relevant class and style attributes, or for LaTeX emitting environments such as `\begin{<class>} ... \end{<class>}`. End users would then only need to provide the relevant CSS or LaTeX definitions via a header include. This would mean that Pandoc wouldn’t have to keep defining and handling special cases, such as centring text.

If I'm not mistaken, that's already happening to some extent.

Parsing: 

```
% pandoc -t native
::: center
foo bar
:::

[ Div
    ( "" , [ "center" ] , [] )
    [ Para [ Str "foo" , Space , Str "bar" ] ]
]
```

Writer HTML: 

```
% pandoc -t html  
::: center
foo bar
:::

<div class="center">
<p>foo bar</p>
</div>
```

Writer LaTeX:

```
% pandoc -t latex
::: center
foo bar
:::

foo bar
```

... The TeX editor needs attention and a (simple?) addition to do without the Lua filters.

**Comment by bpj** [comment]
@cagix wrote:

> Why do we need a special syntax just to center text? Since Pandoc supports fenced divs, couldn't we just use the existing Pandoc Markdown syntax of ::: center ... :::?

That is indeed what I do with filters! One problem with that is that you end up with an empty line between between the `\begin{center}`/`\end{center}` and the content which may or may not affect the output. Another is that you might not like using English words as markup — after all that may be one reason that you don't use LaTeX directly. Sure you can have your filter recognise `centrerat`/`zentriert`(sp?) and what not, but...

However it is hardly feasible to come up with a punctuation markup for every LaTeX environment someone might want to use.I even end up using filters which "convert" custom attributes to other attributes so that I can conveniently use the same source for both LaTeX and HTML: it is far easier to build a `style` attribute from several other attributes than to parse a style attribute to assemble the corresponding LaTeX markup even now when you can use Lpeg/re.

**Comment by igormcoelho** [comment]
I have been using the https://github.com/pandoc-ext/fonts-and-alignment on Lua, and it works fine for LaTeX and Reveal JS...
But I agree that using words, specially english words, does not make much sense for pandoc.
It just occured to me that perhaps something like >, <, >< and <> could be self explaining and language-agnostic?

> Now that Pandoc supports line blocks, here's some syntax that might work nicely for left/center/right -justifying lines:

I kind of agree with @uvtc original idea (just slightly disagree with center x justifying syntax `><` )

For me, the right flush (or raggedleft) could be  as in Lua extension:

```
::: {.flushleft}
Solution using Lua extension
:::
```
Proposed solution:

```
::: <
I don't know if symbol < works as a tag for blocks... at least
it's not language! This makes text on left side, without justifying it.
:::
```

So, the others could be:

```
::: >
                I don't know if symbol > works as a tag for blocks... at least it's not language!
                                            This makes text on right side, without justifying it.
:::
```

```
::: ><
       I don't know if symbol >< works as a tag for blocks... at least it's not language!
                          This makes text on center, without justifying it.
:::
```

```
::: <>
I don't  know if symbol  <>  works  as  a tag  for  blocks...  at  least
it's not language! This makes text on left side, but also justifying it.
:::
```

So, yet another proposal... for over 10 years discussion :)

[EDIT] - I REALLY think this is fundamental for Pandoc, as a filter or some official way... since I didn't know Pandoc AST very well, I took some time with experiments (also for those who feel interested).

<details>

<summary>For Technical Details for Implementation, open this here...</summary>

### Motivation for a clean alternative

It's true that we can use fonts-and-alignment extension, but it uses words in English, and some are hard to remember... also some troubles, like center x centering, flushleft x raggedright, etc.. technical details for LaTeX that won't appear in other generators like JS. So, I'll try to defend `::: >` notation.

### The fenced_divs extension `:::`

When you use ::: you generate a Div in pandoc, and you can add attributes .something (like fenced code attributes)

```
echo '::: { #center .small .other x=1 }    
teste                                                                                                                  
:::' | pandoc -t native 
```
Result is:
```
[ Div
    ( "center" , [ "small" , "other" ] , [ ( "x" , "1" ) ] )
    [ Para [ Str "teste" ] ]
]
```

### Using `::: <` is still Div!

Luckily, this is still a Div with `<` attribute:

```
echo '::: <
teste
:::' | pandoc -t native
```
Result is:
```
[ Div ( "" , [ "<" ] , [] ) [ Para [ Str "teste" ] ] ]
```

However, complicating it with anything else, like `::: < {.param}`, becomes a non-div:
```
echo '::: < {.param}
teste
:::' | pandoc -t native 
```
Result is:
```
[ Para
    [ Str ":::"
    , Space
    , Str "<"
    , Space
    , Str "{.param}"
    , SoftBreak
    , Str "teste"
    , SoftBreak
    , Str ":::"
    ]
]
```

It means that for a filter to keep understanding this as Div, we cannot use more options together. Something like `.<` is quite ugly in my opinion, like `::: {.<}`, so I wouldn't recommend.

### A simple filter in Lua to convert `::: ><` into `.center`

It's easy to do a filter in Lua to convert `::: ><` into .center notation, and then let other existing filters (like fonts-and-alignment) to finish the work.

```
echo '::: ><
   this is centered
:::' | pandoc -t native   --lua-filter=debug-ast.lua   --lua-filter=preprocess-classes.lua   --lua-filter=debug-ast.lua
```
Result is:
```
=== AST RECEIVED ===
[ Div
    ( "" , [ "><" ] , [] )
    [ Para
        [ Str "this" , Space , Str "is" , Space , Str "centered" ]
    ]
]
DEBUG: mapping  '><' → 'center'
=== AST UPDATED  ===
[ Div
    ( "" , [ "center" ] , [] )
    [ Para
        [ Str "this" , Space , Str "is" , Space , Str "centered" ]
    ]
]
```
 
I someone is interested, I can share the source code, but for now only discussing the idea... but it's already usable!

### The case for Justifying `<>` and `><`

At first, I was in need for some Justifying option, but it become more complex for many reasons.
First, it is very hard to manually write thing justified, and it would not honor the visual markdown spirit!
So, now I assume that Justified or Non-justified is simply a global option (that I manually trigger in LaTeX and JS templates) so we can focus here on left align, right align and center. The syntax for center is more naturally `><`, but original author here proposed `<>`... so, I propose to use the one proposed here (`<>`), or perhaps, both.

### The case for simple centerings with `|<`

I think that for single-line centerings, it's better to not use fenced divs, but Line Blocks as proposed here. 
``` 
echo '
| a
| b

teste

|> a
|> b
' | pandoc -t native   --lua-filter=debug-ast.lua   --lua-filter=preprocess-classes.lua   --lua-filter=debug-ast.lua
```
Result is:
```
[ LineBlock [ [ Str "a" ] , [ Str "b" ] ]
, Para [ Str "teste" ]
, Para
    [ Str "|>"
    , Space
    , Str "a"
    , SoftBreak
    , Str "|>"
    , Space
    , Str "b"
    ]
]
```
But this proposal affects more the Pandoc parser, so it's harder (and slightly more aggressive) to do a filter for that, I think (transforming Para into LineBlock).

### Another final topic on columns
A common problem here is also columns, that it's on English and hard to remember the precise notation.

So, ideally for me, columns would also be some variation of fenced divs with `::: ||` (for `.columns`), and then a single one `::: |` (for `.column`).

If somehow the parser incorporates this notation TOGETHER with attributes (I think it's also kind of aggressive), this would allow this kind of structure:

```
:::::::::::::: {.columns align=center totalwidth=8em}
::: {.column width="40%"}
contents...
:::
::: {.column width="60%" align=bottom}
contents...
:::
::::::::::::::
```
written simpler as:
```
:::::::::::::: |><|{totalwidth=8em}
::: |{width="40%"}
contents...
:::
::: |{width="60%" align=bottom}
contents...
:::
::::::::::::::
```
If you glue attributes with pipes, that's fine for now, like this `|><|{totalwidth=8em}` (or perhaps `<>||{totalwidth=8em}`, reading "center columns with totalwidth=8em"), but it makes the whole parsing of attributes a task for the Lua extension... which seems to be kind of trouble.

</details>

**Comment by igormcoelho** [comment]
I finally created some Lua filter to avoid English in my Portuguese texts, and support basic stuff like Centering, change font size, create double-column documents, etc. It's named [div-shortcuts](https://github.com/igormcoelho/pandoc-ext-div-shortcuts) and depends on filter fonts-and-alignment.
I tried my best to keep it within Pandoc philosophy, so I hope it's acceptable. In any case, it's just some filter. Example to use it:

```
pandoc input.md -o output.pdf \
  --pdf-engine=xelatex \
  --lua-filter=div-shortcuts.lua \
  --lua-filter=fonts-and-alignment.lua
```

Examples:

```markdown
::::: >
       This text is aligned to 
                     the right.
:::::

::::: <>
          This text 
              is 
           centered.
:::::
```

```markdown
::::: -
Small text.
:::::

::::: ++++
Huge text.
:::::

```

Also double columns (note that `width` is default attribute for `.column`):

```markdown
:::::::::::::: ||

::::: |{60%}
Left column content.
:::::

::::: |{width=40%}
Right column content.
:::::

::::::::::::::
```

That's it. Yet another proposal for community to discuss more dozens of years :)

**Comment by raffaelemancuso** [comment]
> I finally created some Lua filter to avoid English in my Portuguese texts, and support basic stuff like Centering, change font size, create double-column documents, etc. It's named [div-shortcuts](https://github.com/igormcoelho/pandoc-ext-div-shortcuts) and depends on filter fonts-and-alignment. I tried my best to keep it within Pandoc philosophy, so I hope it's acceptable. In any case, it's just some filter. Example to use it:
> 
> ```
> pandoc input.md -o output.pdf \
>   --pdf-engine=xelatex \
>   --lua-filter=div-shortcuts.lua \
>   --lua-filter=fonts-and-alignment.lua
> ```
> 
> Examples:
> 
> ::::: >
>        This text is aligned to 
>                      the right.
> :::::
> 
> ::::: <>
>           This text 
>               is 
>            centered.
> :::::
> 
> ::::: -
> Small text.
> :::::
> 
> ::::: ++++
> Huge text.
> :::::
> 
> Also double columns (note that `width` is default attribute for `.column`):
> 
> :::::::::::::: ||
> 
> ::::: |{60%}
> Left column content.
> :::::
> 
> ::::: |{width=40%}
> Right column content.
> :::::
> 
> ::::::::::::::
> 
> That's it. Yet another proposal for community to discuss more dozens of years :)

That wouldn't work with Word output, right?

**Comment by igormcoelho** [comment]
> That wouldn't work with Word output, right?

I think it's very doable on Word, yes. The div-shortcuts filter I made does not need any change, since it just translates non-English commands into the existing commands used in fonts-and-alignment filter, which is I guess only made for LaTeX and JS, not Word.
So, the best path to make this combination work (div-shortcuts + fonts-and-alignment) is to propose some Issue on the fonts-and-alignment filter, to see if author is willing to support Word. Since font sizes are relative in LaTeX and absolute in Word, one solution is to propose standardized and proportional changes over base value (example, document is 12pt, so what is Large? +200%?) and then either hardcode it on filter or to somehow describe it in Reference.docx.
The columns part may be trickier for word, since it's based on "::: columns" syntax and I only tested for presentations on Beamer/LaTeX and RevealJS... so perhaps Powerpoint would be easier? I don't know about Word. But it's not covered in fonts-and-alignment either.
When I have some time I can try some prototype of alternative fonts-and-alignment Lua filter in Word... is it a real need or some theoretical discussion? Just to know how motivating this could be :)
